Objects/floatobject.c

```diff
@@ -923,6 +923,52 @@ float___trunc___impl(PyObject *self)
     return PyLong_FromDouble(wholepart);
 }
 
+/*[clinic input]
+float.__floor__
+
+Return the floor of x as an Integral.
+[clinic start generated code]*/
+
+static PyObject *
+float___floor___impl(PyObject *self)
+/*[clinic end generated code: output=e0551dbaea8c01d1 input=38c8fb1dd0b579b4]*/
+{
+    double x = PyFloat_AsDouble(self);
+    double wholepart;
+
+    (void)modf(x, &wholepart);
+    if (LONG_MIN < wholepart && wholepart < LONG_MAX && x < 0) {
+        long aslong = (long)wholepart;
+        if (wholepart != x)
+            --aslong;
+        return PyLong_FromLong(aslong);
+    }
+    return PyLong_FromDouble(wholepart);
+}
+
+/*[clinic input]
+float.__ceil__
+
+Return the ceiling of x as an Integral.
+[clinic start generated code]*/
+
+static PyObject *
+float___ceil___impl(PyObject *self)
+/*[clinic end generated code: output=a2fd8858f73736f9 input=e3d4c94c2b8c8675]*/
+{
+    double x = PyFloat_AsDouble(self);
+    double wholepart;
+
+    (void)modf(x, &wholepart);
+    if (LONG_MIN < wholepart && wholepart < LONG_MAX && x > 0) {
+        long aslong = (long)wholepart;
+        if (wholepart != x)
+            ++aslong;
+        return PyLong_FromLong(aslong);
+    }
+    return PyLong_FromDouble(wholepart);
```

Comment:
This looks wrong: what if the original `x` isn't an exact integer? Then we still need to increment `wholepart` in this branch. (This can't happen with 64-bit longs, where any float with a nonzero fractional part will take the short path, but it can on a machine with 32-bit longs.) When you fix this, please also include a test case.